manifest icons at least {144 192}

paulirish opened the pull request:
Comments welcome.

brendankenny commented on audits/manifest/icons-min-144.js:
```diff
@@ -0,0 +1,69 @@
+   * @return {!AuditResult}
+   */
+  static audit(artifacts) {
+    let hasIconAtLeast144 = false;
```
we should think about a `debugString` for a failure case here. This is a good proving case for `debugString`, actually. Should there be a `debugString` for every possible failure case? (no manifest found, no icons on manifest, etc), or just for the specific case ("No icon at least 144x144 in size found in Manifest" or whatever). Personally I'd lean toward only the specific case, but curious what you think.

paulirish replied:
I like that idea. I'd want to defer to `audits/manifest/icons.js` for the `no icons on manifest` case. can we early exit this 144 check if `icons` false's out?

brendankenny replied:
yeah, makes sense.

brendankenny commented on audits/manifest/icons-min-144.js:
```diff
@@ -0,0 +1,69 @@
+      const sizesArray = icon.value.sizes.value;
+      !!sizesArray && sizesArray.forEach(size => {
+        const pair = size.split(/x/i);
+        if (pair[0] && parseFloat(pair[0]) >= 144) {
```
chrome tests both numbers. Is there any requirement an icon be square? I've never seen a reference to one that wasn't...

brendankenny replied:
FirefoxOS requires a square, and Chrome app docs say "you should ensure that all of your icons are square, or unexpected behavior may result.", but maybe we should just check width and height to be thorough.

paulirish replied:
when looking at the source yesterday it assume a square. and it'll distort if the image isn't a square. So we could debugstring if we find its not square, but we can't fail 'em since thing still work.

brendankenny replied:
Sorry, I meant we shouldn't check that they're equal, but we should check that both are ≥ 144

brendankenny commented on audits/manifest/icons-min-144.js:
```diff
@@ -0,0 +1,69 @@
+      return ManifestIconsMin144.generateAuditResult(false);
+    }
+
+    manifest.icons.value.forEach(icon => {
```
to early exit the search when a valid icon is found, what about switching `forEach()` to `find()`, and inside the first return the result of the nested `find()` call and return the `parseFloat` conditional inside that?

brendankenny commented:
needs tests! You can probably just copy over the icons-192 tests, with an extra test to get the > 144 conditional.

Also, if Sam were here, he'd ask you to squash :)

Looking good, though. Do you want to modify icons-192.js in this same manner?

paulirish commented:
Okay one big refactor later…

* Tests added. 
* `icons-at-least` helper introduced. much cleaner logic
* Now both 144 and 192 test use it
* debugString for unequal sizes added at the parser layer.
* Did I mention tests? :)

PTAL

brendankenny commented on closure/closure-type-checking.js:
```diff
@@ -28,6 +28,7 @@ gulp.task('js-compile', function() {
     'closure/typedefs/*.js',
     'closure/third_party/*.js',
     'audits/**/*.js',
+    'helpers/icons-at-least.js',
```
nice :)

brendankenny commented on helpers/manifest-parser.js:
```diff
@@ -141,8 +141,13 @@ function parseIcon(raw) {
+    sizes.value.split(/\s/).forEach(size => {
+      const pair = size.split(/x/i);
+      if (pair[0] !== pair[1]) {
+        sizes.debugString = 'WARNING: icon sizes are not equal, expect the unexpected!';
```
haha, this makes it sound fun

samccone replied:
:+1:

brendankenny commented on helpers/manifest-parser.js:
```diff
@@ -141,8 +141,13 @@ function parseIcon(raw) {
-
+    sizes.value.split(/\s/).forEach(size => {
+      const pair = size.split(/x/i);
+      if (pair[0] !== pair[1]) {
```
my only question on this is if we want this to live in the parser. Just a warning, so it may be fine, but non-square icons are a-ok as far as the spec is concerned.

brendankenny replied:
also, where should this warning be displayed? A particular audit? We don't currently have a place where general manifest weirdness (but non-errors) can be logged

paullewis replied:
I think we should surface errors with the results in both the CLI and extension, under each line item.

brendankenny replied:
right, but this is being generated in a gather (manifest.js), so it needs a particular audit to decided to display it. Currently this will just be dropped.

paulirish replied:
IMO i think the check belongs here in the parser. I'd be in favor of exposing the debugString for any icon-size check audits, for the winning pair. 

e.g. if theres 3 icons: (92x128, 144x144, 192x192), then the user never sees the debugString.

seem good, brendan?

brendankenny replied:
That sounds good for debugString.

My concern for where it lives is just mixing our layers here. Sam had mentioned making it so others could easily use our manifest parser, and having things that are browser-specific and aren't in the spec (like, things get weird if your icons aren't square) feel like they should live on code that builds on the parser, not in it.

paullewis replied:
Agreed. We want to be agnostic here. The manifest gathering should be about parsing the manifest, and the audit is where we sense-check the values that we've got. Whether those values actually matter is for the aggregator to decide.

brendankenny commented on audits/manifest/icons-min-144.js:
```diff
@@ -0,0 +1,63 @@
+  static audit(artifacts) {
+    if (iconsAudit.audit(artifacts).value === false) {
+      return ManifestIconsMin144.generateAuditResult(false, undefined,
+        'WARNING: No icons found in the manifest');
```
I thought you wanted to just have a false in this case and let the icons exist audit handle the `debugString`?

paulirish replied:
true!

We shouldn't necc show the user the has-icons audit. It's an implementation detail that supports the 192 and 144 tests. 

Using that approach in my next patch

brendankenny commented on helpers/icons-at-least.js:
```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+/**
+ * @param  {Number} sizePx
```
`number`

brendankenny replied:
we also don't have an eslint check for this (yet), but I usually see just one space after `@param`

brendankenny commented on helpers/icons-at-least.js:
```diff
@@ -0,0 +1,34 @@
+
+/**
+ * @param  {Number} sizePx
+ * @param  {any} manifest
```
`{*}` (I'm surprised this one didn't throw an error...)

brendankenny replied:
oh, actually you want `{!ManifestNode<(!Manifest|undefined)>}` (the type of `artifacts.manifest`). *But*, since you already know at the call sites that there's a `manifest.value`, you could pass in `artifacts.manifest.value` instead and type it as a simple `{!Manifest}`

brendankenny commented on helpers/icons-at-least.js:
```diff
@@ -0,0 +1,34 @@
+ */
+module.exports = function iconsAtLeast(sizePx, manifest) {
+  const iconValues = manifest.value.icons;
+  const nestedSizes = iconValues.value.map(icon => icon.value.sizes.value);
```
it might make sense to do a filter to toss any `icon.value.sizes.value` that are undefined here, then you can avoid the `typeof` check below

brendankenny commented on helpers/icons-at-least.js:
```diff
@@ -0,0 +1,34 @@
+      .map(size => size.split(/x/i))
+      .map(pairStr => [parseFloat(pairStr[0]), parseFloat(pairStr[1])])
+      .filter(pair => pair[0] >= sizePx)
+      .filter(pair => pair[1] >= sizePx);
```
this seems...a little excessive. At least combine the size comparisons into one :P

paullewis replied:
Add comments for each step of that plz :smile:

brendankenny commented on test/audits/manifest/icons-min-144.js:
```diff
@@ -0,0 +1,81 @@
+        sizes: '96x96 128x128 144x144 256x256'
+      }]
+    });
+    const manifest = manifestParser(manifestSrc);
```
maybe a manifest with two icons, one without any sizes and the other with a valid size?

brendankenny replied:
and maybe a non-square size? Just to follow the lead of https://code.google.com/p/chromium/codesearch#chromium/src/chrome/browser/banners/app_banner_data_fetcher_unittest.cc&sq=package:chromium&type=cs&l=105

paulirish replied:
sg. added both.

paullewis commented on audits/manifest/icons-min-144.js:
```diff
@@ -0,0 +1,63 @@
+   */
+  static audit(artifacts) {
+    if (iconsAudit.audit(artifacts).value === false) {
+      return ManifestIconsMin144.generateAuditResult(false, undefined,
```
This looks decidedly "boolean-trap"-ish. Wondering if we should be making `generateAuditResult` take an object.

paulirish replied:
Im in favor of that. Can we do that, @brendankenny ?

brendankenny replied:
makes sense, though I'm not wild about `generateAuditResult` in general. We'll be passing an object to a function just for that function to return the same object with a name and description added, which it got from our own class :)

paulirish replied:
Aye. Let's take this discussion to #135

paullewis commented on audits/manifest/icons-min-192.js:
```diff
@@ -31,14 +33,14 @@ class ManifestIcons192 extends Audit {
    */
   static get name() {
-    return 'manifest-icons-192';
+    return 'manifest-icons-min192';
```
`min-192`*

paullewis commented on test/manifest-parser-tests.js:
```diff
@@ -55,6 +56,29 @@ describe('Manifest Parser', function() {
   });
 
+  describe('icon parsing', function() {
+    it('it parses basic string', function() {
```
double 'it'

paullewis commented on test/manifest-parser-tests.js:
```diff
@@ -55,6 +56,29 @@ describe('Manifest Parser', function() {
+      assert.equal(parsedManifest.value.icons.value.length, 1);
+    });
+
+    it('it warns if unequal sizes are presented', function() {
```
Same.

paullewis commented on test/manifest-parser-tests.js:
```diff
@@ -55,6 +56,29 @@ describe('Manifest Parser', function() {
+      assert.equal(parsedManifest.value.icons.value.length, 1);
+    });
+
+    it('it finds three icons in the stub manifest', function() {
```
Saaaame :D

paullewis commented:
Lots going on here, but def the right kinds of stuff.

paulirish commented:
Pushed new code:

* Icons-exist shouldn't be shown to the user, so it's moved from an audit into a helper
* Single test file for both 144 and 192 audits, since it's all the same stuff.
* Added documentation and refactored the `icons-at-least` stuff.

ptal

paullewis commented on audits/manifest/icons-min-144.js:
```diff
@@ -46,19 +48,19 @@ class ManifestIcons192 extends Audit {
-      });
-      hasIcons = (!!icons192);
+    if (iconsExist(manifest) === false) {
+      return ManifestIconsMin144.generateAuditResult(false, undefined,
```
Thinking I should deal with this in #138

paullewis commented on audits/manifest/icons-min-144.js:
```diff
@@ -46,19 +48,19 @@ class ManifestIcons192 extends Audit {
-    return ManifestIcons192.generateAuditResult(hasIcons);
+    const matchingIcons = iconsAtLeast(144, manifest);
+
+    return ManifestIconsMin144.generateAuditResult(!!matchingIcons.length, undefined,
```
Same.

paullewis commented:
I think we need to be a bit stricter on where the logic lives (as mentioned in the code comments), but other than that I'd be happy to LGTM.

samccone commented on audits/manifest/icons-min-192.js:
```diff
@@ -18,8 +18,10 @@
 'use strict';
 
 const Audit = require('../audit');
+const iconsExist = require('../../helpers/icons-exist');
```
can we merge the icon helpers into just a helpers/icons.js file?

paulirish replied:
great call. doing it.

samccone commented on helpers/icons-at-least.js:
```diff
@@ -0,0 +1,41 @@
+      .filter(size => typeof size === 'string')
+      .filter(size => {
+        // Split the '24x24' strings into ['24','24'] arrays
+        const sizeStrs = size.split(/x/i);
```
I feel like this regex can be smarter to look for numbers on either side of the x

brendankenny replied:
that string is generated by the manifest-parser, so should be guaranteed to be in that format or not in the parsed output.

samccone replied:
kk :+1: we should note that in the closure docs for manifestValue.icons

paulirish replied:
so you're thinking we do one of these first?

  if (/\d+x\d+/.test(size) === false) // exit with error

samccone commented on helpers/icons-exist.js:
```diff
@@ -0,0 +1,33 @@
+/**
+ * @param {!Manifest} manifest
+ */
+module.exports = function iconsExist(manifest) {
```
see previous comment about merging the icon helpers into a single file

samccone commented:
Not to be pedantic I but I would really like to have tests around the icon helpers :)

brendankenny commented:
:+1:

paulirish commented:
guys you're going to LOVE this, but before it lands, bckenny has to debug some weird closure error.

samccone commented:
this needs to be squashed as well

brendankenny commented on helpers/icons.js:
```diff
@@ -0,0 +1,67 @@
+'use strict';
+
+/**
+ * @param {!Manifest=} manifest
```
`@return {boolean}`

brendankenny commented on helpers/icons.js:
```diff
@@ -0,0 +1,67 @@
+/**
+ * @param {number} sizeRequirement
+ * @param {!Manifest} manifest
+ */
```
`@return {!Array<boolean>}`

brendankenny replied:
oh right, `filter`, so `@return {!Array<string>}` :)

brendankenny commented on helpers/manifest-parser.js:
```diff
@@ -139,8 +139,7 @@ function parseIcon(raw) {
     let set = new Set();
-    sizes.value.split(/\s/).forEach(size => set.add(size.toLowerCase()));
-
+    sizes.value.trim().split(/\s+/).forEach(size => set.add(size.toLowerCase()));
```
good catch

brendankenny replied:
https://developers.whatwg.org/common-microsyntaxes.html#set-of-space-separated-tokens

paulirish replied:
\o/

brendankenny replied:
should it be trimmed, though? That's not clear to me from the spec. If so, you should actually set `let sizes = parseString(raw.sizes, true);` up above to match the other cases where strings are trimmed, then have `sizes.value.split(/\s+/).forEach(size => set.add(size.toLowerCase()));` here

brendankenny replied:
I can't think of any reasonable interpretation of this that wouldn't trim first and [Chrome appears to agree](https://code.google.com/p/chromium/codesearch#chromium/src/third_party/WebKit/Source/core/html/HTMLLinkElementSizesAttributeTest.cpp&sq=package:chromium&type=cs&l=35) :)

brendankenny commented on test/manifest-parser-tests.js:
```diff
@@ -55,14 +56,31 @@ describe('Manifest Parser', function() {
   });
 
+  describe('icon parsing', function() {
+    it('parses basic string', function() {
```
add some intermediate variables here? It's not at all clear what you're testing once you get a level or two deep

paulirish replied:
done

brendankenny commented:
looking good, a few last things

brendankenny commented on test/manifest-parser-tests.js:
```diff
@@ -55,14 +56,31 @@ describe('Manifest Parser', function() {
+      assert(!parsedManifest.value.icons.value[0].value.sizes.debugString);
+      assert.equal(parsedManifest.value.icons.value.length, 3);
+    });
+  });
```
add a test for the case you added (icons with multiple spaces in the sizes string)?

paulirish replied:
k

paulirish commented:
Updated

brendankenny commented:
LGTM!

paulirish commented:
Cheers. thanks yall. <3
